src/Npgsql/Internal/PgTypeInfoResolverChainBuilder.cs

```diff
@@ -0,0 +1,129 @@
+using System;
+using System.Collections.Generic;
+
+namespace Npgsql.Internal;
+
+struct PgTypeInfoResolverChainBuilder
+{
+    readonly List<PgTypeInfoResolverFactory> _factories = new();
+    Action<PgTypeInfoResolverChainBuilder, List<IPgTypeInfoResolver>>? _addRangeResolvers;
+    Action<PgTypeInfoResolverChainBuilder, List<IPgTypeInfoResolver>>? _addMultirangeResolvers;
+    RangeArrayHandler _rangeArrayHandler = RangeArrayHandler.Instance;
+    MultirangeArrayHandler _multirangeArrayHandler = MultirangeArrayHandler.Instance;
+    Action<PgTypeInfoResolverChainBuilder, List<IPgTypeInfoResolver>>? _arrayResolvers;
+
+    public PgTypeInfoResolverChainBuilder()
+    {
+    }
+
+    public void Clear() => _factories.Clear();
+
+    public void AppendResolverFactory(PgTypeInfoResolverFactory factory) => AddResolverFactory(factory);
+    public void PrependResolverFactory(PgTypeInfoResolverFactory factory) => AddResolverFactory(factory, prepend: true);
+
+    void AddResolverFactory(PgTypeInfoResolverFactory factory, bool prepend = false)
+    {
+        var type = factory.GetType();
+
+        for (var i = 0; i < _factories.Count; i++)
+            if (_factories[i].GetType() == type)
+            {
+                _factories.RemoveAt(i);
+                break;
+            }
+
+        if (prepend)
+            _factories.Insert(0, factory);
+        else
+            _factories.Add(factory);
+    }
+
+    public void EnableRanges()
+    {
+        _addRangeResolvers ??= AddResolvers;
+        _rangeArrayHandler = RangeArrayHandlerImpl.Instance;
+
+        static void AddResolvers(PgTypeInfoResolverChainBuilder instance, List<IPgTypeInfoResolver> resolvers)
+        {
+            foreach (var factory in instance._factories)
+                if (factory.CreateRangeResolver() is { } resolver)
+                    resolvers.Add(resolver);
+        }
+    }
+
+    public void EnableMultiranges()
+    {
+        _addMultirangeResolvers ??= AddResolvers;
+        _multirangeArrayHandler = MultirangeArrayHandlerImpl.Instance;
+
+        static void AddResolvers(PgTypeInfoResolverChainBuilder instance, List<IPgTypeInfoResolver> resolvers)
+        {
+            foreach (var factory in instance._factories)
+                if (factory.CreateMultirangeResolver() is { } resolver)
+                    resolvers.Add(resolver);
+        }
+    }
+
+    public void EnableArrays()
+    {
+        _arrayResolvers ??= AddResolvers;
+
+        static void AddResolvers(PgTypeInfoResolverChainBuilder instance, List<IPgTypeInfoResolver> resolvers)
+        {
+            foreach (var factory in instance._factories)
+                if (factory.CreateArrayResolver() is { } resolver)
+                    resolvers.Add(resolver);
+
+            if (instance._addRangeResolvers is not null)
```

Comment:
There's a part of me that wants to kick out range arrays by default (on slim), since they're unlikely to be useful (or what the user wants) given that multiranges are a thing.

On the other hand, this is unlikely to be significant in terms of size savings (there are only a few ranges anyway), this already happens only if the user enabled both arrays and ranges, and it's probably a good idea to keep the invariant that if you've enabled arrays, then every base type has its array etc.

Reply:
It should be very insignificant. Since multiranges already root most of the array and list types including these range array mappings means we're paying for some instantiations of ArrayBasedArrayConverter/ListBasedArrayConverter (both of which are ~1kb each as far as I recall).

Reply:
>Since multiranges already root most of the array and list types
Of course that presupposes the user also called EnableMultiranges. 

If it ever comes up - I strongly doubt it - we can always expose the internal array resolver granularity via more specific opt-ins on the builder as well. We always keep the option to expose full control to the user.

Reply:
Nah...